### DISCUSSION TITLE
Boolean as a model input

### DISCUSSION THREAD

**Discussion by salehisaeed** [serizba/cppflow#208] [Q&A] [answered]
Hi all,

I have a basic question regarding loading a trained model using CppFlow.

So, my multi-input model should be called through:
`    auto output = model
    (
        {
            {"serving_default_args_0:0", states},
            {"serving_default_deterministic:0", deterministic}
        },
        {"StatefulPartitionedCall:0"}
    );
`
The `states` is my float inputs that can be created through `cppflow::fill({1, 99}, 1.0f);`. But the `deterministic` input is a boolean. I am not sure how to produce it. I tried many possibilities without success.

Any idea or help would be much appreciated.
Saeed

**Comment by serizba** [comment, marked as answer]
Hi @salehisaeed 

Perhaps the best way to get a bool tensor is to use cast:

```c++
unsigned char mytensor = 0;
auto c = cppflow::tensor(mytensor);
cppflow::cast(mytensor, TF_UINT8, TF_BOOL);
```

**Reply by tdakhran** [reply]
@serizba , thanks for the suggestion, it works!
I tried to specialize constructor of `cppflow::tensor` for `bool`. I created an intermediate `uint8_t` vector, then copied boolean data into it, and then called the private constructor with `TF_BOOL`, but got ```what():  Invalid handle```.

**Reply by serizba** [reply]
Could you post the piece of code you tried?

**Reply by tdakhran** [reply]
This is what I tried
```cpp
template<>
tensor::tensor(const std::vector<bool>& values, const std::vector<int64_t>& shape) 
{
    std::vector<uint8_t> data(begin(values), end(values));
    tensor(TF_BOOL, data.data(), values.size() * sizeof(bool), shape) ;
}
```
and getting
```bash
terminate called after throwing an instance of 'std::runtime_error'
  what():  Invalid handle
```

But I found the issue, need to assign `*this = tensor(TF_BOOL, data.data(), values.size() * sizeof(bool), shape) ;` and it fixes the problem.

**Reply by serizba** [reply]
Exactly! Glad that you fixed it!

**Reply by salehisaeed** [reply]
Hi @serizba 

Thanks for your reply. The `c` tensor is not used in your proposed code:
> ```c++
> unsigned char mytensor = 0;
> auto c = cppflow::tensor(mytensor);
> cppflow::cast(mytensor, TF_UINT8, TF_BOOL);
> ```

Did you mean?
```c++
 cppflow::cast(c, TF_UINT8, TF_BOOL);
```
Nevertheless, it still does not work for me.  The `c` tensor still has a type of `dtype=TF_UINT8` after casting and thus my model fails.

**Reply by salehisaeed** [reply]
I tried @tdakhran suggestion and added a specialized constructor to `cppflow::tensor` and it works OK. 

Thank you all for the help.

**Reply by serizba** [reply]
Probably the @tdakhran solution is better.

What I meant was:

```c++
unsigned char c = 0;
auto mytensor = cppflow::tensor(c);
mytensor = cppflow::cast(mytensor, TF_UINT8, TF_BOOL);
```
Perhaps you didn't assign the casted tensor. `cppflow::cast` returns a tensor, so you need to assign it.

**Reply by salehisaeed** [reply]
Correct! Now, this solution also works.

Thanks!

**Comment by tdakhran** [comment]
@serizba, does it make sense to upstream the support of bool tensors with the overhead of extra copy? I can prepare PR.
